### ISSUE TITLE
Request: Working grpc server + client example

### ISSUE THREAD

**Issue by dapirian** [protobufjs/protobuf.js#1598]
In README.md there's some example code of making a grpc client, but nothing for making a server, and there's seemingly nothing in the docs or examples explaining how to make a server. Is there any example using protobuf.js that could be pointed to for this?

**Comment by alexander-fenster** [comment]
@dapirian The official gRPC [guide](https://grpc.io/docs/languages/node/basics/) actually uses protobuf.js (via `@grpc/proto-loader`).  The `protoLoader.loadSync(...)` call in the guide actually translates to a `loadSync()` call from protobuf.js.

**Comment by dapirian** [comment]
> @dapirian The official gRPC [guide](https://grpc.io/docs/languages/node/basics/) actually uses protobuf.js (via `@grpc/proto-loader`). The `protoLoader.loadSync(...)` call in the guide actually translates to a `loadSync()` call from protobuf.js.

I see, so I'm using `pbjs` + `pbts` to generate `.js` + `.d.ts` from a proto with a service and some message definitions in it and trying to use those files directly. Is there anything you can point to about that?

**Comment by alexander-fenster** [comment]
OK so Bazel is a different story. In protobufjs 6.x, `pbjs` is non-hermetic (it does `npm install` the first time it's run) which makes plugging it in Bazel build much harder than it should be. I tried and gave up, so I just decided to commit the generated files for the time being.  Good news is that in the upcoming protobufjs 7.x the CLI tools are moved to their separate package and the runtime dependency problem will be fixed, which will hopefully make it easy to wrap that into a `genrule` or something.

Just FYI, I'm maintaining the Bazel [project](https://github.com/googleapis/gapic-generator-typescript) written in TypeScript with protos.

Now back to your question, there are two options:

1. Make the gRPC code read the `*.proto` files, parsing them at runtime using `@grpc/proto-loader` (the `.loadSync()` method). In this case, you won't need `pbjs` or `pbts`.
2. Make the gRPC code read the JSON file generated by `pbjs -t json -o protos.json *.proto` (the `.loadSync()` method will happily accept a JSON filename instead of a proto filename).
3. Make the gRPC code read the parsed JSON file generated by `pbjs -t json -o protos.json *.proto` (you can use `.fromJSON()` method of `@grpc/proto-loader`, passing the JSON content (`require("./protos.json")`) to it. This option will be the most friendly to bundlers should you want to use them.

To summarize that - there is no way to make gRPC use the generated `*.js` / `*.d.ts` files, you would need to generate a JSON representation if you want to use `pbjs` and `pbts`.

Hope all this makes sense :)

**Comment by dapirian** [comment]
Thanks for the info! I have `pbjs`/`pbts` running in bazel via [these bazel rules](https://github.com/bazelbuild/rules_nodejs/blob/stable/examples/protobufjs/defs.bzl) (might be of interest to you). Like you said, until the cli tools are their own package it's a pain. I think I'll probably do option #3 from above, but I'm not sure about the typescript types. Looks like `proto-loader` has `proto-loader-gen-types.js` which you can use to generate typescript type info. I could wrap calling that in a bazel rule along with `pbjs` generating the json. I'm a little confused about the ecosystem between `proto-loader` and protobuf.js` though... I'm unsure how the generated types from `pbts` would differ from `proto-loader-gen-types.js` (I'll try it out and see though)

**Comment by alexander-fenster** [comment]
The TypeScript types in the proto loader is a new feature in 0.6.x, I haven't tried it so can't tell you much about it. @murgatroid99 implemented it and can give you more details :) I believe the two sets of types might be similar but differ in details.

In Google Cloud libraries we want to export types for gRPC clients, and we have a post-processing step for `pbts` generated types [here](https://github.com/googleapis/gax-nodejs/blob/master/tools/compileProtos.ts), but this is of course a hack. This all started long before the proto-loader implemented its own type generation.

**Comment by dcodeIO** [comment]
Closing this issue because it has been inactive for more than two years and the project has changed substantially.

If this is still relevant, please open a fresh issue with current reproduction steps, version details, and any updated context.

Please also consider that maintaining protobuf.js takes time and effort. If the project is valuable to you or your organization, sponsorship is appreciated.

Thanks!
